pycloudlib/qemu/cloud.py

```diff
@@ -0,0 +1,634 @@
+        subprocess.run(
+            [
+                "qemu-img",
+                "create",
+                "-f",
+                "qcow2",
+                "-b",
+                base_image,
+                "-F",
+                "qcow2",
+                str(instance_path),
+                "10G",  # This space is only taken when needed
+            ],
+            check=True,
+        )
+
+    def _get_ubuntu_kernel_from_image_dir(self, image: Path) -> Optional[Path]:
+        kernels = list(
+            image.parent.glob("*-server-cloudimg-amd64-vmlinuz-generic")
+        )
+        if len(kernels) != 1:
+            self._log.warning("Unable to find kernel for image: %s", image)
+            return None
+        return kernels[0].absolute()
+
+    def _find_base_image(self, image_id: str) -> Path:
+        if Path(image_id).exists():
+            base_image = Path(image_id)
+        elif Path(self.image_dir, image_id).exists():
+            base_image = Path(self.image_dir, image_id)
+        else:
+            raise ImageNotFoundError(
+                resource_id=image_id,
+                message=(
+                    f"Could not find '{image_id}' as absolute path "
+                    f"or in '{self.image_dir}'."
+                ),
+            )
+        self._log.debug("Using base image: %s", base_image)
+        return base_image
+
+    def _parse_instance_type(self, instance_type: str) -> Tuple[int, int]:
+        cpu_mem = re.search(r"^c(?P<cpus>\d+)m(?P<memory>\d+$)", instance_type)
+        if not cpu_mem:
+            raise ValueError(
+                "instance_type must be in the form of c#m#, not "
+                f"{instance_type}"
+            )
+        return int(cpu_mem["cpus"]), int(cpu_mem["memory"])
+
+    def _get_kernel_path(
+        self, kernel_path, kernel_cmdline: str, image_id, base_image
+    ) -> Optional[Path]:
+        kernel = None
+        if kernel_cmdline and not kernel_path:
+            kernel = self._get_ubuntu_kernel_from_image_dir(image=base_image)
+            if not kernel:
+                raise PycloudlibError(
+                    f"Could not find kernel for image_id: {image_id}. "
+                    "Please specify kernel_path."
+                )
+        return kernel
+
+    def _update_kernel_cmdline(self, kernel_cmdline: str) -> str:
+        default_kernel_args = ""
+        if "root=" not in kernel_cmdline:
+            default_kernel_args += "root=/dev/sda1 "
+        if "console=" not in kernel_cmdline:
+            default_kernel_args += "console=ttyS0 "
+        return default_kernel_args + kernel_cmdline
+
+    # pylint:disable=R0914
+    def launch(
+        self,
+        image_id: str,
+        instance_type="c1m1500",
+        user_data=None,
+        meta_data=None,
+        vendor_data=None,
+        kernel_cmdline="",
+        kernel_path=None,
+        no_seed_iso=False,
+        username: Optional[str] = None,
+        launch_args: Optional[List[str]] = None,
```

Comment:
I do think it would be helpful to provide a supplmental `qemu_args: Optional[List[str]]`  to extend or manipulate the default qemu behavior. I know there are cases for the desktop images where I'd want to deal with `-daemonize` or `--nographic` depending on my test usecases.

Reply:
Do you mean something more than [this](https://github.com/canonical/pycloudlib/pull/323/files#diff-9c031a07867513f1651aa8c522f0ca475eed8b5756bafa4ebaa92966811fdb57R528). I do realize that won't let you unwrite/rewrite certain parameters, but I'm not sure the best way to support that.